### DISCUSSION TITLE
Localization - Supported locales not accepting custom values and missing from official Cultural languages and others

### DISCUSSION THREAD

**Discussion by mpalavrov** [keycloak/keycloak#23554] [Archive]
Below I will mention few things about the Locales (bugs or improvements, not sure how to categorize):
### 1. Cannot add custom Locales via the Admin Console UI with the keycloak.v2 theme
**en-CA** for example
![image](https://github.com/keycloak/keycloak/assets/33352511/d6cb08b2-2586-4114-a157-7b001a07d462)

For work-around the REST API can be used with PUT request (after GET request to get the full Realm object) with the new **supportedLocales**, like:
```
    "supportedLocales": [
        "en", "de","en-ZA", "en-AU", "TEST"
    ],
```

### 2. A lot of locales are missing from the list of the pre-defined ones.
Currently, there are 144 added
![image](https://github.com/keycloak/keycloak/assets/33352511/01d5d16f-38d9-4400-abe4-f4707abb9cd4)
`العربية
català
Deutsch
Ελληνικά
English
español
فارسی
français
suomi
italiano
日本語
lietuvių
latviešu
Nederlands
norsk
polski
português (Brasil)
русский
svenska
中文（中国）
čeština
dansk
magyar
slovenčina
Türkçe
Afar
Abkhazian
Afrikaans
አማርኛ
অসমীয়া
Aymara
azərbaycan
Bashkir
беларуская
български
भोजपुरी
Bislama
বাংলা
བོད་སྐད་
brezhoneg
Corsican
Cymraeg
རྫོང་ཁ
esperanto
eesti
euskara
Fijian
føroyskt
Frysk
Gaeilge
Gàidhlig
galego
Guarani
ગુજરાતી
Hausa
עברית
हिन्दी
hrvatski
հայերեն
interlingua
Indonesia
Interlingue
Inupiaq
Indonesia
íslenska
Inuktitut
עברית
ייִדיש
Jawa
ქართული
қазақ тілі
kalaallisut
ខ្មែរ
ಕನ್ನಡ
한국어
کٲشُر
kurdî
кыргызча
Latin
lingála
ລາວ
Malagasy
Māori
македонски
മലയാളം
монгол
română
मराठी
Melayu
Malti
မြန်မာ
Nauru
नेपाली
Occitan
Oromoo
ଓଡ଼ିଆ
ਪੰਜਾਬੀ
پښتو
português
Runasimi
rumantsch
Ikirundi
română
Kinyarwanda
संस्कृत भाषा
سنڌي
Sängö
srpski (latinica)
සිංහල
slovenščina
Samoan
chiShona
Soomaali
shqip
српски
Swati
Southern Sotho
Basa Sunda
Kiswahili
தமிழ்
తెలుగు
тоҷикӣ
ไทย
ትግርኛ
türkmen dili
Filipino
Tswana
lea fakatonga
Tsonga
татар
Akan
ئۇيغۇرچە
українська
اردو
o‘zbek
Tiếng Việt
Volapük
Wolof
IsiXhosa
ייִדיש
Èdè Yorùbá
Zhuang
中文
isiZulu`

More complete list of locales, **228**, (including cultural ones as well), you can find in https://gist.github.com/JamieMason/3748498. I have prepared a list as well, which includes more locales, **297**, combined for multiple sources:
`af
af-ZA
am-ET
ar
ar-AE
ar-BH
ar-DZ
ar-EG
ar-IQ
ar-JO
ar-KW
ar-LB
ar-LY
ar-MA
arn-CL
ar-OM
ar-QA
ar-SA
ar-SD
ar-SY
ar-TN
ar-YE
as-IN
az
az-az
az-Cyrl-AZ
az-Latn-AZ
ba-RU
be
be-BY
bg
bg-BG
bn-BD
bn-IN
bo-CN
br-FR
bs-Cyrl-BA
bs-Latn-BA
ca
ca-ES
co-FR
cs
cs-CZ
cy-GB
da
da-DK
de
de-AT
de-CH
de-DE
de-LI
de-LU
dsb-DE
dv
dv-MV
el
el-CY
el-GR
en
en-029
en-AU
en-BZ
en-CA
en-cb
en-GB
en-IE
en-IN
en-JM
en-MT
en-MY
en-NZ
en-PH
en-SG
en-TT
en-US
en-ZA
en-ZW
es
es-AR
es-BO
es-CL
es-CO
es-CR
es-DO
es-EC
es-ES
es-GT
es-HN
es-MX
es-NI
es-PA
es-PE
es-PR
es-PY
es-SV
es-US
es-UY
es-VE
et
et-EE
eu
eu-ES
fa
fa-IR
fi
fi-FI
fil-PH
fo
fo-FO
fr
fr-BE
fr-CA
fr-CH
fr-FR
fr-LU
fr-MC
fy-NL
ga-IE
gd-GB
gd-ie
gl
gl-ES
gsw-FR
gu
gu-IN
ha-Latn-NG
he
he-IL
hi
hi-IN
hr
hr-BA
hr-HR
hsb-DE
hu
hu-HU
hy
hy-AM
id
id-ID
ig-NG
ii-CN
in-ID
is
is-IS
it
it-CH
it-IT
iu-Cans-CA
iu-Latn-CA
iw-IL
ja
ja-JP
ka
ka-GE
kk
kk-KZ
kl-GL
km-KH
kn
kn-IN
ko
kok
kok-IN
ko-KR
ky
ky-KG
lb-LU
lo-LA
lt
lt-LT
lv
lv-LV
mi-NZ
mk
mk-MK
ml-IN
mn
mn-MN
mn-Mong-CN
moh-CA
mr
mr-IN
ms
ms-BN
ms-MY
mt-MT
nb-NO
ne-NP
nl
nl-BE
nl-NL
nn-NO
no
no-no
nso-ZA
ns-ZA
oc-FR
or-IN
pa
pa-IN
pl
pl-PL
prs-AF
ps-AF
pt
pt-BR
pt-PT
qut-GT
quz-BO
quz-EC
quz-PE
rm-CH
ro
ro-mo
ro-RO
ru
ru-mo
ru-RU
rw-RW
sa
sah-RU
sa-IN
se-FI
se-NO
se-SE
si-LK
sk
sk-SK
sl
sl-SI
sma-NO
sma-SE
smj-NO
smj-SE
smn-FI
sms-FI
sq
sq-AL
sr
sr-BA
sr-CS
sr-Cyrl-BA
sr-Cyrl-CS
sr-Cyrl-ME
sr-Cyrl-RS
sr-Latn-BA
sr-Latn-CS
sr-Latn-ME
sr-Latn-RS
sr-ME
sr-RS
sr-sp
sv
sv-FI
sv-SE
sw
sw-KE
syr
syr-SY
ta
ta-IN
te
te-IN
tg-Cyrl-TJ
th
th-TH
tk-TM
tlh-QS
tn-ZA
tr
tr-TR
tt
tt-RU
tzm-Latn-DZ
ug-CN
uk
uk-UA
ur
ur-PK
uz
uz-Cyrl-UZ
uz-Latn-UZ
uz-uz
vi
vi-VN
wo-SN
xh-ZA
yo-NG
zh-CHS
zh-CHT
zh-CN
zh-HK
zh-MO
zh-SG
zh-TW
zu-ZA`

even if the locales cannot be added via the Admin Console UI, once added via the REST API, some of them, for example like **en-AU**, **en-ZA**, etc. are recognized in the **Edit message bundles** section:
![image](https://github.com/keycloak/keycloak/assets/33352511/9e93aaa4-7ebe-4ff6-a227-42090214db1f)
but still, not addable in the Supported Locales section and there they are shown with their code only:
![image](https://github.com/keycloak/keycloak/assets/33352511/cffc6565-2844-48ad-8960-7b11f7b6782d)

### 3. The locales to be searchable in the "Supported Locales" section using their locale key, like:

- English to be searchable with **en** and **English**
- Bulgarian to be searchable with **bg**, **Bulgarian** and **български** (which is the local language naming)
- Arabic to be searchable with **ar**, **Arabic** and its local language name
same apply to all languages/locales

### 5. Translations in the UI is not working for custom added locale (TEST for example):
![image](https://github.com/keycloak/keycloak/assets/33352511/c0b48095-3b56-4500-8f60-c487213507cb)
Even if the above locale_TEST key is set to some custom value, when selected, it is not used, but for the translation in the Drop-Down it is falling back to the English translation and for selected language in the drop-down you can see "TEST", but in lower case:
![image](https://github.com/keycloak/keycloak/assets/33352511/ad5f7a57-5d5c-42c1-a3e7-afe0f94b6010)
The same is for every translation for custom locale (similar to **TEST**)

### 6. Have consistency between the "keycloak.v2" Admin Account Console languages translations in the "Edit message bundles" section and the "keycloak" Login theme.
Currently in the Login theme you can see this translations:
![image](https://github.com/keycloak/keycloak/assets/33352511/5b49d7e7-f966-4e21-8f42-8accbcbf5942)
and in the Admin Account Console:
![image](https://github.com/keycloak/keycloak/assets/33352511/a6a3ff96-16ab-457e-90b2-227d09ebdded)
where the **TEST** locale is having the **locale_TEST** key:
![image](https://github.com/keycloak/keycloak/assets/33352511/47fa7f98-dca1-4fcd-a6d9-ee62e8daaeee)

I can imagine that having some consistency in the locales/languages names translation would be better, than having it differently.

**Comment by thodorispa** [comment]
Hi did you manage to resolve it?

**Comment by bdovaz** [comment]
Issue created: https://github.com/keycloak/keycloak/issues/36859

I hope they fix it
